packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,116 @@
+      associationValue.set(Constraints.ASSOCIATION, association);
+      this.associationMap.set(associationKey, associationValue);
+    }
+  }
+
+  public static addResourceUpdateConstraint(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct, options: TagUpdatesOptions) {
+    const associationKey = this.associationPrecheck(scope, portfolio, product);
+    InputValidator.validateLength(this.prettyPrintAssociation(scope, portfolio, product), 'description', 0, 2000, options.description);
+    if (!this.associationMap.get(associationKey)!.has(Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.acceptedMessageLanguage,
+        description: options.description,
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+        tagUpdateOnProvisionedProduct: options.allowUpdatingProvisionedProductTags === false ? Allowed.NOT_ALLOWED : Allowed.ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey)!.get(Constraints.ASSOCIATION)!);
+      this.associationMap.get(associationKey)!.set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.prettyPrintAssociation(scope, portfolio, product)}`);
+    }
+  }
+
+  public static associateTagOption(scope: cdk.Resource, resourceId: string, tagOption: TagOption) {
+    Object.keys(tagOption).forEach(key => {
+      InputValidator.validateLength(resourceId, 'TagOption key', 1, 128, key);
+      tagOption[key].forEach(value => {
+        InputValidator.validateLength(resourceId, 'TagOption value', 1, 256, value.value);
+        const tagOptionKey = hashValues(key, value.value, scope.stack.node.addr);
+        if (!this.tagOptionMap.has(tagOptionKey)) {
+          const cfnTagOption = new CfnTagOption(scope, `TagOption${tagOptionKey}`, {
+            key: key,
+            value: value.value,
+            active: value.active ?? true,
+          });
+          this.tagOptionMap.set(tagOptionKey, cfnTagOption);
+
+          new CfnTagOptionAssociation(scope, `TagOptionAssociation${hashValues(scope.node.addr, tagOptionKey)}`, {
+            resourceId: resourceId,
+            tagOptionId: this.tagOptionMap.get(tagOptionKey)!.ref,
+          });
+        }
+      });
+    });
+  }
+
+  private static associationMap = new Map<string, Map<Constraints, CfnConstraintResource>>();
+  private static tagOptionMap = new Map<string, CfnTagOption>();
+
+  private static associationPrecheck(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): string {
+    const associationKey = hashValues(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      this.associateProductWithPortfolio(scope, portfolio, product);
```

Comment:
Can't you just call this method always? It's safe to call even if the product and the portfolio are already associated, right? It does its own check inside of it.

Reply:
Still a valid question I think.

Reply:
we certainly could, I feel like it sacrifices some readability though.  Unless the method name is changed to reflect that it also checks.

Reply:
Sorry, what readability does it sacrifice?

Reply:
before it is clear that it checks if association exists, and then if not will try to create it,  if we get rid of it then it's not clear in this method, they would have to go to called method to see it still checks.  But I ended up going with recommendation and adding a comment 
```
private static ensureProductIsAssociatedWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): string {
    const associationKey = this.getAssociationKey(scope, portfolio, product);
    // This method will check if the association already exists before creating a new one
    this.associateProductWithPortfolio(scope, portfolio, product);
    return associationKey;
  }
```